doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,796 @@
+for a socket becoming readable/writeable and then call `SSL_read()`/`SSL_write()`
+or some other `SSL` function in response. For servers it can also "accept" new
+connections on a socket to learn about new connection events.
+
+With QUIC there will be a single socket that is shared between multiple `SSL`
+objects (or possibly a set of sockets that are shared). A socket becoming
+readable/writeable can no longer be associated with a single `SSL` object. New
+connections are just new packets arriving on the socket(s) just like any other.
+There will need to be a mechanism to deliver events to an application that are
+associated with an individual `SSL` object in order for the standard `SSL_read()`/
+`SSL_write(`) functions, or some other IO function to be called.
+
+To solve this problem this design introduces a new `SSL_EVENT_CTX` object. An
+individual `SSL` object represents an individual stream. A set of `SSL` objects
+together represent all of the streams available for a given connection. The
+first `SSL` object created for a connection represents the connection itelf as
+well as the first stream if there is one. All the `SSL` objects from all
+connections associated with a shared set of sockets are all grouped together by
+a single `SSL_EVENT_CTX`.
+
+While the principal motivation for introducing `SSL_EVENT_CTX` is to support QUIC,
+it may also be useful for supporting (D)TLS applications. Using an `SSL_EVENT_CTX`
+will be required for QUIC applications, but using one should (if desired by the
+application developer) also be possible for (D)TLS applications. It should also
+be possible to write (D)TLS applications in the traditional way.
+
+OpenSSL (D)TLS applications abstract away the details of the underlying
+transport protocol using `BIO`s. Application developers can either use one of the
+set of standard `BIO`s, or alternatively develop their own custom `BIO`s in order
+to have more control over the interaction with the underlying transport
+protocol. This approach needs to be possible with QUIC too. However for QUIC it
+will no longer be the case that each `SSL` object needs different `BIO` objects.
+Instead there will be common `BIO`s that are shared between multiple `SSL` objects.
+These shared `BIO`s can be set on the `SSL_EVENT_CTX`.
+
+Applications can call `SSL_EVENT_CTX_get_next_event()` to get the next event from
+the event queue. Examples of events might be "new connection", "new stream",
+"stream readable", "stream close", etc. Time based events can also be delivered.
+Optionally events can be filtered for only a given `SSL` object. If the event
+queue is empty then a new packet will be read from the underlying `BIO`s (if
+possible) and any resulting events are added to the event queue.
+
+Also associated with the `SSL_EVENT_CTX` is a new connection queue and a new
+stream queue. In the case that a "new connection" event is received then the
+application would need to create a new SSL object and then call `SSL_accept()` on
```

Comment:
Can be created in the normal way `SSL_new(ctx)`. As long as it has the correct `SSL_EVENT_CTX` set for it (either inherited from the `SSL_CTX` or set explicitly then it should all just work.

*Possibly* it could be created automatically - but I was trying to keep things as close as possible to the way things are currently done. It would not be difficult to add that capability later if we wanted.

Reply:
I am slightly worried if we do not constrain ourselves with this design into similar problems the current DTLSv1_listen/SSL_accept has with losing packets.

But yeah I suppose this is not set in stone and we can adjust the API if we find out later.